src/nvim/sign.c

```diff
@@ -612,206 +193,139 @@ static linenr_T buf_delsign(buf_T *buf, linenr_T atlnum, int id, char *group)
-    if (sign->se_id == id && sign_in_group(sign, group)) {
-      return (int)sign->se_lnum;
-    }
-  }
-
-  return 0;
+  uint32_t ns_id = (uint32_t)map_get(String, int)(&namespace_ids, cstr_as_string(group));
+  return (id < 0 || (ns_id == 0 && group != NULL)) ? 0
+         : marktree_lookup_ns(buf->b_marktree, ns_id, (uint32_t)id, false, NULL).pos.row + 1;
 }
 
-/// Return the sign at line 'lnum' in buffer 'buf'. Returns NULL if a sign is
-/// not found at the line. If 'groupname' is NULL, searches in the global group.
+/// Delete the specified signs
 ///
-/// @param buf  buffer whose sign we are searching for
-/// @param lnum  line number of sign
-/// @param groupname  sign group name
-static sign_entry_T *buf_getsign_at_line(buf_T *buf, linenr_T lnum, char *groupname)
+/// @param buf  buffer sign is stored in or NULL for all buffers
+/// @param group  sign group
+/// @param id  sign id
+/// @param atlnum  sign at this line, -1 at any line
+static int buf_delete_signs(buf_T *buf, char *group, int id, linenr_T atlnum)
 {
-  sign_entry_T *sign;    // a sign in the signlist
+  int64_t ns_id = group_get_ns(group);
+  if (ns_id < 0) {
+    return FAIL;
+  }
 
-  FOR_ALL_SIGNS_IN_BUF(buf, sign) {
-    if (sign->se_lnum > lnum) {
-      // Signs are sorted by line number in the buffer. No need to check
-      // for signs after the specified line number 'lnum'.
-      break;
+  MarkTreeIter itr[1];
+  int row = atlnum > 0 ? atlnum - 1 : 0;
+  kvec_t(MTKey) signs = KV_INITIAL_VALUE;
+
+  if (atlnum > 0) {
+    if (!marktree_itr_get_overlap(buf->b_marktree, row, 0, itr)) {
+      return FAIL;
     }
 
-    if (sign->se_lnum == lnum && sign_in_group(sign, groupname)) {
-      return sign;
+    MTPair pair;
+    while (marktree_itr_step_overlap(buf->b_marktree, itr, &pair)) {
+      if ((ns_id == UINT32_MAX || ns_id == pair.start.ns)
+          && pair.start.decor_full && decor_has_sign(pair.start.decor_full)) {
+        kv_push(signs, pair.start);
+      }
     }
+  } else {
+    marktree_itr_get(buf->b_marktree, 0, 0, itr);
   }
 
-  return NULL;
-}
-
-/// Return the identifier of the sign at line number 'lnum' in buffer 'buf'.
-///
-/// @param buf  buffer whose sign we are searching for
-/// @param lnum  line number of sign
-/// @param groupname  sign group name
-static int buf_findsign_id(buf_T *buf, linenr_T lnum, char *groupname)
-{
-  sign_entry_T *sign;   // a sign in the signlist
-
-  sign = buf_getsign_at_line(buf, lnum, groupname);
-  if (sign != NULL) {
-    return sign->se_id;
+  while (itr->x) {
+    MTKey mark = marktree_itr_current(itr);
+    if (row && mark.pos.row > row) {
+      break;
+    }
+    if ((id == 0 || (int)mark.id == id)
+        && (ns_id == UINT32_MAX || ns_id == mark.ns)
+        && !mt_end(mark) && mark.decor_full && decor_has_sign(mark.decor_full)) {
+      if (atlnum > 0) {
+        kv_push(signs, mark);
+        marktree_itr_next(buf->b_marktree, itr);
+      } else {
+        extmark_del(buf, itr, mark, 0, 0);
```

Comment:
Is this `extmark_del()` refactor alright and if so is there any reason why it shouldn't be used in `extmark_clear()`? I noticed that adds all marks/decors to a vector instead of deleting during iteration.